protocol/bc/vmcontext.go

```diff
@@ -0,0 +1,89 @@
+package bc
+
+import (
+	"bytes"
+
+	"chain/protocol/vm"
+)
+
+func NewBlockVMContext(block *Block, prog []byte, args [][]byte) *vm.Context {
+	blockHash := block.Hash().Bytes()
+	return &vm.Context{
+		VMVersion: 1,
+		Code:      prog,
+		Arguments: args,
+
+		BlockHash:            &blockHash,
+		BlockTimeMS:          &block.TimestampMS,
+		NextConsensusProgram: &block.ConsensusProgram,
+	}
+}
+
+func NewTxVMContext(tx *Tx, index uint32, prog Program, args [][]byte) *vm.Context {
+	var (
+		txSigHash        = tx.SigHash(index).Bytes()
+		numResults       = uint64(len(tx.Results))
+		assetID          = tx.Inputs[index].AssetID()
+		assetIDBytes     = assetID[:]
+		amount           = tx.Inputs[index].Amount()
+		inputRefDataHash = hashData(tx.Inputs[index].ReferenceData).Bytes()
+		txRefDataHash    = hashData(tx.ReferenceData).Bytes()
+	)
+
+	checkOutput := func(index uint64, refdatahash []byte, amount uint64, assetID []byte, vmVersion uint64, code []byte) (bool, error) {
+		if index >= uint64(len(tx.Outputs)) {
+			return false, vm.ErrBadValue
```

Comment:
Did you have thoughts on my actual comment above?

> But since it can't [be a simple predicate], it might look a little cleaner to have it return `error` that's one of `vm.ErrBadValue` or `vm.NotFound` or `nil`.